linera-views/src/shared_view.rs

```diff
@@ -0,0 +1,220 @@
+}
+
+impl<C, V> SharedView<C, V>
+where
+    V: View<C>,
+{
+    /// Wraps a `view` in a [`SharedView`].
+    pub fn new(view: V) -> Self {
+        SharedView {
+            view,
+            reader_count: Arc::new(AtomicUsize::new(0)),
+            writer_semaphore: Arc::new(Semaphore::new(1)),
+            writer_notifier: Arc::new(Mutex::new(None)),
+            _context: PhantomData,
+        }
+    }
+
+    /// Returns a [`ReadOnlyViewReference`] to the inner [`View`].
+    ///
+    /// If there is a writer with a [`ReadWriteViewReference`] to the inner [`View`], waits
+    /// until that writer is finished.
+    pub async fn inner(&mut self) -> Result<ReadOnlyViewReference<V>, ViewError> {
+        let _no_writer_check = self.writer_semaphore.acquire().await;
+
+        self.reader_count.fetch_add(1, Ordering::SeqCst);
+
+        Ok(ReadOnlyViewReference {
+            view: self.view.share_unchecked()?,
+            reader_count: self.reader_count.clone(),
+            writer_notifier: self.writer_notifier.clone(),
+        })
+    }
+
+    /// Returns a [`ReadWriteViewReference`] to the inner [`View`].
+    ///
+    /// Waits until the previous writer is finished if there is one. There can only be one
+    /// [`ReadWriteViewReference`] to the same inner [`View`].
+    pub async fn inner_mut(&mut self) -> Result<ReadWriteViewReference<V>, ViewError> {
+        let writer_permit = self.writer_semaphore.acquire_arc().await;
+        let (sender, receiver) = oneshot::channel();
+        let mut writer_notifier = self
+            .writer_notifier
+            .lock()
+            .expect("No panics should happen while holding the `write_notifier` lock");
+
+        if let Some(notifier) = writer_notifier.take() {
+            assert!(
+                notifier.send(()).is_err(),
+                "`writer_notifier` should either be already used or the receiving writer should \
+                have already finished"
+            );
+        }
+
+        *writer_notifier = Some(sender);
+
+        Ok(ReadWriteViewReference {
+            view: self.view.share_unchecked()?,
+            write_barrier: Some(receiver),
+            _writer_permit: writer_permit,
+        })
+    }
+}
+
+/// A read-only reference to a [`SharedView`].
+pub struct ReadOnlyViewReference<V> {
+    view: V,
+    reader_count: Arc<AtomicUsize>,
+    writer_notifier: Arc<Mutex<Option<oneshot::Sender<()>>>>,
+}
+
+impl<V> Deref for ReadOnlyViewReference<V> {
+    type Target = V;
+
+    fn deref(&self) -> &Self::Target {
+        &self.view
+    }
+}
+
+impl<V> Drop for ReadOnlyViewReference<V> {
+    fn drop(&mut self) {
+        if self.reader_count.fetch_sub(1, Ordering::SeqCst) == 1 {
+            let mut writer_notifier = self
+                .writer_notifier
+                .lock()
+                .expect("No panics should happen while holding the `write_notifier` lock");
+
+            if self.reader_count.load(Ordering::Acquire) == 0 {
```

Comment:
Would it be possible to put the `let mu write_notifier` into the `if self.reader_count`?